### PR TITLE
[i2c, test] Add i2c test for communicating with a compass device

### PR DESCRIPTION
**Pull request by jwnrt**
This PR adds an i2c host test for communicating with a compass device. The device is asked to take a measurement which is read back by the host.

Also includes cosmetic fixes to `i2c_host_power_monitor_test` which I copied to make this test.

### CODE REVIEW

File: sw/device/tests/i2c_host_compass_test.c [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+  TRY(dif_i2c_init(base_addr, &i2c));
+
+  base_addr = mmio_region_from_addr(TOP_EARLGREY_PINMUX_AON_BASE_ADDR);
+  TRY(dif_pinmux_init(base_addr, &pinmux));
+
+  TRY(i2c_testutils_connect_i2c_to_pinmux_pins(&pinmux, 0));
+
+  TRY(dif_i2c_host_set_enabled(&i2c, kDifToggleEnabled));
+  // Enable the external IRQ at Ibex.
+  irq_global_ctrl(true);
```
Comment (engdoreis) [inline sw/device/tests/i2c_host_compass_test.c:108]: These two lines should be removed as you may not use interrupts in this test.

**Review by engdoreis — APPROVED** [review]
Thanks for the PR @jwnrt.

File: sw/device/tests/i2c_host_compass_test.c (hunk 1 above)
Comment (jwnrt) [reply sw/device/tests/i2c_host_compass_test.c]: Oh thanks!

**Comment by jwnrt** [comment]
Removed the irq stuff and fixed an unhelpful error message.

**Comment by jwnrt** [comment]
Aside: is there a way to retry kokoro without force pushing? For now I'll just rebase.

**Comment by andreaskurth** [comment]
> Aside: is there a way to retry kokoro without force pushing? For now I'll just rebase.

Yes, you can add the `kokoro:rebuild` label, which will get automatically removed once kokoro has been retriggered.

Adding @marnovandermaas for I2C HW

File: sw/device/tests/BUILD [hunk 2]
```diff
@@ -1786,7 +1786,7 @@ opentitan_functest(
     name = "i2c_host_power_monitor_test",
     srcs = ["i2c_host_power_monitor_test.c"],
     cw310 = cw310_params(
-        tags = ["skip_in_ci"],  # Requires the Bob in CI.
+        tags = ["manual"],  # Requires the Bob in CI.
```
Comment (marnovandermaas) [inline sw/device/tests/BUILD]: Does this Bob also need to be BoB?

File: sw/device/tests/BUILD [hunk 3]
```diff
@@ -1806,6 +1806,28 @@ opentitan_functest(
     ],
 )
 
+opentitan_functest(
+    name = "i2c_host_compass_test",
+    srcs = ["i2c_host_compass_test.c"],
+    cw310 = cw310_params(
+        tags = ["skip_in_ci"],  # Requires the Bob in CI.
```
Comment (marnovandermaas) [inline sw/device/tests/BUILD:1813]: Does this Bob need to be BoB?

**Review by marnovandermaas — APPROVED** [review]
LGTM, just some small nits.

File: sw/device/tests/BUILD (hunk 3 above)
Comment (jwnrt) [reply sw/device/tests/BUILD:1835]: Ah yes, thanks

Comment (jwnrt) [reply sw/device/tests/BUILD]: It should also have been `"manual"`. Fixed.

**Comment by jwnrt** [comment]
Rebased to fix conflicts, but more arrived in between so sorry for all the force pushes!

I've also now included a commit to fix a `CHECK` in the `EXECUTE_TEST` macro which fails for tests that overflow 32b cycles.

The first time I ran this test after fixing it, the first read gave back all `0x00`. I've seen this happen with the accelerometer, too, but haven't been able to reproduce afterwards.

Maybe this was due to leaving the i2c in a bad state from previous failed runs, and then the first i2c command on the next run fails? That's the best guess I have.

When we enable these in CI it would be nice to mark them as non-blocking if they fail or something similar. I'm not sure we have anything like that in Bazel, though.

File: sw/device/lib/testing/test_framework/ottf_main.h [hunk 4]
```diff
@@ -116,9 +116,9 @@ char *ottf_task_get_self_name(void);
     uint64_t t_start_ = ibex_mcycle_read();                              \
     status_t local_status = INTO_STATUS(test_function_());               \
     uint64_t t_end_ = ibex_mcycle_read();                                \
-    CHECK(t_end_ <= UINT32_MAX);                                         \
-    CHECK(t_start_ <= t_end_);                                           \
-    uint32_t cycles_ = (uint32_t)t_end_ - (uint32_t)t_start_;            \
+    uint64_t cycles64_ = t_end_ - t_start_;                              \
```
Comment (engdoreis) [inline sw/device/lib/testing/test_framework/ottf_main.h:119]: Sorry for not noticing this before. I'm suggesting some minor changes to make this macro smaller.
```suggestion
    uint64_t cycles_ = ibex_mcycle_read() - t_start_;                              \
```

File: sw/device/lib/testing/test_framework/ottf_main.h [hunk 5]
```diff
@@ -116,9 +116,9 @@ char *ottf_task_get_self_name(void);
...
+    CHECK(cycles64_ <= UINT32_MAX);                                      \
```
Comment (engdoreis) [inline sw/device/lib/testing/test_framework/ottf_main.h:120]:
```suggestion
    CHECK(cycles_ <= UINT32_MAX);                                      \
```

Comment (jwnrt) [reply sw/device/lib/testing/test_framework/ottf_main.h]: Thanks, updated with the shorter macro

FYI GitHub lets you select multiple line numbers, then you can do a suggestion over all of them in one comment

**Review by marnovandermaas — APPROVED** [review]
LGTM, just a small nit from my side.

File: sw/device/lib/testing/test_framework/ottf_main.h [hunk 6]
```diff
@@ -115,13 +115,11 @@ char *ottf_task_get_self_name(void);
     LOG_INFO("Starting test " #test_function_ "...");                    \
     uint64_t t_start_ = ibex_mcycle_read();                              \
     status_t local_status = INTO_STATUS(test_function_());               \
-    uint64_t t_end_ = ibex_mcycle_read();                                \
-    CHECK(t_end_ <= UINT32_MAX);                                         \
-    CHECK(t_start_ <= t_end_);                                           \
```
[the remark this answers is not in the document]
Comment (marnovandermaas) [reply sw/device/lib/testing/test_framework/ottf_main.h:120]: Yes, it would be a huge overflow. That's true, so probably fine to leave it as is then.

**Pull request merged**